hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/SystemContractAbis.java

```diff
@@ -0,0 +1,176 @@
+        this.version = version;
+        this.signature = stripBlanks(signature);
+        this.decoderSignature = stripName(addressToUint(this.signature));
+        this.decoder = TypeFactory.create(this.decoderSignature);
+        this.selector = Bytes.wrap(new Function(this.signature).selector());
+        this.methodName = nameFrom(this.signature);
+    }
+
+    SystemContractAbis(@NonNull final String signature) {
+        this(1, signature);
+    }
+
+    public final int version;
+    public final String signature;
+    public final String decoderSignature;
+    public final ABIType<Tuple> decoder;
+    public final Bytes selector;
+    public final String methodName;
+
+    @NonNull
+    public String selectorAsHex() {
+        return to8CharHexString(selector);
+    }
+
+    @NonNull
+    String nameFrom(@NonNull final String s) {
+        // signature must look like `fooMethod(arg1,arg2,arg3)` so this just pulls out that
+        // part before the first `(`
+        return s.substring(0, s.indexOf('('));
+    }
+
+    @NonNull
+    String addressToUint(@NonNull final String s) {
+        // for some reason the decoder library doesn't accept the Solidity type `address` and
+        // requires the equivalent `bytes32` instead
+        return s.replace("address", "bytes32");
+    }
+
+    @NonNull
+    String stripBlanks(@NonNull final String s) {
+        return s.replaceAll("\\s+", "");
+    }
+
+    @NonNull
+    String stripName(@NonNull final String s) {
+        // signature must look like `fooMethod(arg1,arg2,arg3)` so this just trims off
+        // the `fooMethod` at the beginning
+        return s.substring(s.indexOf('('));
+    }
+
+    @NonNull
+    String to8CharHexString(@NonNull final Bytes bytes) {
+        // `Bytes` has a bunch of "toHex"-ish routines - but not one that can be asked to cough
+        // up a _fixed-width_ hex value.  So you've got to pad it yourself.  And the way it is
+        // done here is to glue the hex digits onto the back of 8 `0`s and then just take the last
+        // 8 characters of that.
+        final var shortHex = "00000000" + bytes.toUnprefixedHexString();
+        return "0x" + shortHex.substring(shortHex.length() - 8);
+    }
+
+    /** Safely convert from either a BigInteger or a Long to a Long.
+     * <p>
+     * The method parameter decoder returns Objects. In the case of a Solidity `uint64` it will
+     * be a `BigInteger`, for a `int64` it will be a `Long`.  Safely convert to a `long` (the given
+     * `BigInteger.longValue` does a narrowing primitive conversion (which simply truncates to the
+     * low 64 bits)).
+     */
+    public static long toLongSafely(@NonNull final Object bigNumeric) {
+        if (bigNumeric instanceof BigInteger big) {
+            if (big.compareTo(LONG_MIN) < 0 || big.compareTo(LONG_MAX) > 0)
+                throw new IllegalArgumentException("uint64 out of range for Java.long");
```

Comment:
Do we need to throw or return Long.MIN_VALUE or Long.MAX_VALUE , since its casting to Long

Reply:
This is an shim for fixing the mismatch between the EVMs support of `uint64` and the JVMs lack of it.  So these values are not good and should revert the contract.  Hence: throw.

In fact this is the reason we've got multiple contract methods that have been versioned from V1 to V2 (or beyond): They were initially specified with `uint64` which doesn't work right.  So _newer_ ABIs have signed only.  But we have to support the older ones forever (because: blockchain).

Reply:
(Resolved to enable merging but I think this answer will suffice - if not - let me know!)